src/classes/RD2_NamingService.cls

```diff
@@ -0,0 +1,247 @@
+    */
+    @TestVisible
+    private static RD2_Settings settings {
+        get {
+            if (settings == null) {
+                settings = new RD2_Settings();
+            }
+            return settings;
+        }
+        set;
+    }
+
+    /***
+    * @description Returns Donor (Account/Contact) records
+    */
+    @TestVisible
+    private RD2_Donor.IGateway donorGateway {
+        get {
+            if (donorGateway == null) {
+                donorGateway = new RD2_Donor.Gateway();
+            }
+            return donorGateway;
+        }
+        set;
+    }
+
+    /***
+    * @description Returns Recurring Donation records
+    */
+    @TestVisible
+    private RD2_RecurringDonationsGateway rdGateway {
+        get {
+            if (rdGateway == null) {
+                rdGateway = new RD2_RecurringDonationsGateway();
+            }
+            return rdGateway;
+        }
+        set;
+    }
+
+    /***
+    * @description Determines if multi-currency is enabled for the org
+    */
+    private Boolean isMultiCurrencyEnabled {
+        get {
+            if (isMultiCurrencyEnabled == null) {
+                isMultiCurrencyEnabled = UserInfo.isMultiCurrencyOrganization();
+            }
+            return isMultiCurrencyEnabled;
+        }
+        set;
+    }
+
+    /***
+    * @description Currency symbol for the user default currency
+    */
+    private String userDefaultCurrencySymbol {
+        get {
+            if (userDefaultCurrencySymbol == null) {
+                userDefaultCurrencySymbol = UTIL_Currency.getInstance().getCurrencySymbol(UserInfo.getDefaultCurrency());
+            }
+            return userDefaultCurrencySymbol;
+        }
+        set;
+    }
+
+    /***
+    * @description Autogenerates recurring donation names when Account name has changed
+    */
+    public void updateOnAccountChange(List<Id> accountIds) {
+        if (!settings.isAutomaticNamingEnabled()) {
+            return;
+        }
+
+        List<npe03__Recurring_Donation__c> rds = new List<npe03__Recurring_Donation__c>();
+
+        for (npe03__Recurring_Donation__c rd : rdGateway.getRecordsReferencingAccounts(accountIds)) {
+            rd.Name = System.Label.npo02.NameReplacementText;
+            rds.add(rd);
+        }
+
+        if (!rds.isEmpty()) {
+            UTIL_DMLService.updateRecords(rds);
+        }
+    }
+
+    /***
+    * @description Autogenerates recurring donation names when Contact name has changed
+    */
+    public void updateOnContactChange(Set<Id> contactIds) {
+        if (!settings.isAutomaticNamingEnabled()) {
+            return;
+        }
+
+        List<npe03__Recurring_Donation__c> rds = new List<npe03__Recurring_Donation__c>();
+
+        for (npe03__Recurring_Donation__c rd : rdGateway.getRecordsReferencingContacts(contactIds)) {
+            rd.Name = System.Label.npo02.NameReplacementText;
+            rds.add(rd);
+        }
+
+        if (!rds.isEmpty()) {
```

Comment:
Is there a need to actually push the change to the Name through as DML? Couldn't this method just call the `autogenerateNames()` method with the list of RD's to rename? That would save on DML and processing time.